### PR TITLE
Avoid re-fetching binaries when they exist

### LINKED ISSUE #491: Avoid repeated downloads for assets during test
After the successful integration of #474, our GitHub CI/CD pipeline now automatically publishes a tarball of the latest RISC-V executables, keeping the rv32emu source tree clean. However, we have noticed that the tarball is being downloaded multiple times during tests, leading to unnecessary delays. To improve this, we can cache the downloaded files beforehand and simply verify the extracted RISC-V executables during the testing phase.

### PR DESCRIPTION
**Pull request by vacantron**
Now, we fetch the SHA1 first and verify the binaries. Only if the check fails, the download begins.

### CODE REVIEW

**Review by jserv — CHANGES_REQUESTED** [review]
Improve git commit messages.

If you tend to trigger tests, specify "WIP:" in the subject.

**Comment by ChinYikMing** [comment]
Rather than checking for the existence of prebuilt executable filenames, I prefer calculating and comparing their hash values with pre-calculated ones. For verifying the hash of a file or directory, see the `verify` function in "mk/external.mk".

https://github.com/sysprog21/rv32emu/blob/abb11352c1ab391e67cd94324a83f94d7e8caf4d/mk/external.mk#L47

**Comment by vacantron** [comment]
> Rather than checking for the existence of prebuilt executable filenames, I prefer calculating and comparing their hash values with pre-calculated ones. For calculating the hash of a directory, see the `verify` function in "mk/external.mk".
> 
> https://github.com/sysprog21/rv32emu/blob/abb11352c1ab391e67cd94324a83f94d7e8caf4d/mk/external.mk#L47

This macro is hard to be reused in the recipe. I prefer to write a new shell script to handle the SHA1 verification.

**Review by ChinYikMing — COMMENTED** [review]
Enhance the git commit message by providing more details, specifically explaining how the SHA1 value is handled.

File: mk/artifact.mk [hunk 1]
```diff
@@ -48,12 +48,27 @@ else
...
+	$(Q)$(eval RES := 0)
+	$(Q)$(eval $(foreach FILE,$(PREBUILT_X86_FILENAME), \
+	    $(call verify,$(shell grep -w $(FILE) $(BIN_DIR)/sha1sum-linux-x86-softfp | awk '{ print $$1 };'),$(BIN_DIR)/linux-x86-softfp/$(FILE),RES) \
+	))
+	$(Q)$(eval $(foreach FILE,$(PREBUILT_RV32_FILENAME), \
+	    $(call verify,$(shell grep -w $(FILE) $(BIN_DIR)/sha1sum-riscv32 | awk '{ print $$1 };'),$(BIN_DIR)/riscv32/$(FILE),RES) \
+	))
+
+	$(Q)if [ "$(RES)" = "1" ]; then \
+	    $(PRINTF) "SHA1 verification failed! Re-fetching prebuilt binaries from \"rv32emu-prebuilt\" ...\n"; \
```
Comment (ChinYikMing) [inline mk/artifact.mk]: Add some color to this message, such as using red.

define red color in `mk/common.mk`
```
RED = \033[31m
```

Consider this: 
```
$(PRINTF) "$(RED)SHA1 verification failed! Re-fetching prebuilt binaries from \"rv32emu-prebuilt\" ...\n$(NO_COLOR)"
```

File: mk/artifact.mk [hunk 2]
```diff
@@ -48,12 +48,27 @@ else
...
+.PHONY: artifact fetch-sha1 scimark2 ieeelib
 
-artifact: ieeelib scimark2
+artifact: fetch-sha1 ieeelib scimark2
 ifeq ($(call has, PREBUILT), 1)
-	$(Q)$(PRINTF) "Fetching prebuilt executables from \"rv32emu-prebuilt\" ...\n"
-	$(Q)wget -q --show-progress https://github.com/sysprog21/rv32emu-prebuilt/releases/download/$(LATEST_RELEASE)/rv32emu-prebuilt.tar.gz -O- | tar -C build --strip-components=1 -xz
+	$(Q)$(PRINTF) "Checking SHA1 of binaries ...\n"
+
+	$(eval PREBUILT_X86_FILENAME := $(shell cat $(BIN_DIR)/sha1sum-linux-x86-softfp | awk '{ for(i=1;i<=NF;i++) if(i%2==0) print $$i };'))
```
Comment (ChinYikMing) [inline mk/artifact.mk]: Simplify to:
```
$(eval PREBUILT_X86_FILENAME := $(shell cat $(BIN_DIR)/sha1sum-linux-x86-softfp | awk '{ print $$2 };'))
```

File: mk/artifact.mk [hunk 3]
```diff
@@ -48,12 +48,27 @@ else
...
+	$(Q)$(eval PREBUILT_RV32_FILENAME := $(shell cat $(BIN_DIR)/sha1sum-riscv32 | awk '{ for(i=1;i<=NF;i++) if(i%2==0) print $$i };'))
```
Comment (ChinYikMing) [inline mk/artifact.mk]: Ditto.

**Comment by ChinYikMing** [comment]
Is there any relationship between the 'sha1sums.csv' files in [this link](https://github.com/sysprog21/rv32emu-prebuilt/blob/master/testbenches/x86_64/sha1sums.csv) and [this link](https://github.com/sysprog21/rv32emu-prebuilt/blob/master/testbenches/riscv32/sha1sums.csv) and this commit?

If not, they can be removed from the `rv32emu-prebuilt` repo to keep it clean.

File: mk/external.mk [hunk 4]
```diff
@@ -58,10 +59,13 @@ define verify
                 | cut -f 1 -d ' ' > $(SHA1_FILE2) && cmp $(SHA1_FILE1) $(SHA1_FILE2))), \
             ($(eval VERIFIER := echo "$(strip $(1))  $(strip $(2))" | $(SHA1SUM) -c)) \
     ))
-    $(eval _ := $(shell $(VERIFIER)))
+    $(eval _ := $(shell $(VERIFIER) 2>&1))
```
Comment (ChinYikMing) [inline mk/external.mk:62]: Why make this changes?

File: mk/artifact.mk (hunk 2 above)
Comment (ChinYikMing) [reply mk/artifact.mk]: Append `$(Q)` for consistency.

File: mk/external.mk (hunk 4 above)
Comment (vacantron) [reply mk/external.mk:62]: This redirects `stderr` to `stdout`. The output after running `make artifact` will be in a mess if every SHA1 verifications fail without redirecting error message.

**Comment by vacantron** [comment]
> Is there any relationship between the 'sha1sums.csv' files in [this link](https://github.com/sysprog21/rv32emu-prebuilt/blob/master/testbenches/x86_64/sha1sums.csv) and [this link](https://github.com/sysprog21/rv32emu-prebuilt/blob/master/testbenches/riscv32/sha1sums.csv) and this commit?
> 
> If not, they can be removed from the `rv32emu-prebuilt` repo to keep it clean.

`rv32emu-prebuilt` now only releases the tarball of the prebuilt binaries, and all other files in the repository will not be updated in the future.

File: mk/external.mk (hunk 4 above)
Comment (ChinYikMing) [reply mk/external.mk:62]: > This redirects `stderr` to `stdout`. The output after running `make artifact` will be in a mess if every SHA1 verifications fail without redirecting error message.

Got it.

**Comment by ChinYikMing** [comment]
> > Is there any relationship between the 'sha1sums.csv' files in [this link](https://github.com/sysprog21/rv32emu-prebuilt/blob/master/testbenches/x86_64/sha1sums.csv) and [this link](https://github.com/sysprog21/rv32emu-prebuilt/blob/master/testbenches/riscv32/sha1sums.csv) and this commit?
> > If not, they can be removed from the `rv32emu-prebuilt` repo to keep it clean.
> 
> `rv32emu-prebuilt` now only releases the tarball of the prebuilt binaries, and all other files in the repository will not be updated in the future.

I’m still confused about file suffixes related to CSV. Could you please provide more details?

**Comment by vacantron** [comment]
> I’m still confused about file suffixes related to CSV. Could you please provide more details?

The contents of `*.csv` files are copied from the output of `sha1sum` and are changed to comma-separated manually (the released `sha1sum-*` are still whitespace separated).

File: mk/artifact.mk [hunk 5]
```diff
@@ -82,6 +99,16 @@ else
 	          -DCMAKE_BUILD_TYPE=RELEASE -DBOARD_NAME=rv32emu .. && \
 	    make
 	$(Q)cp ./tests/quake/build/port/boards/rv32emu/quake $(BIN_DIR)/riscv32/quake
+
+	$(Q)(cd $(BIN_DIR)/linux-x86-softfp; for fd in *; do $(SHA1SUM) "$$fd"; done) >> $(BIN_DIR)/sha1sum-linux-x86-softfp
+	$(Q)(cd $(BIN_DIR)/riscv32; for fd in *; do $(SHA1SUM) "$$fd"; done) >> $(BIN_DIR)/sha1sum-riscv32
+endif
+
+fetch-sha1:
```
Comment (jserv) [inline mk/artifact.mk]: Rename to `fetch-checksum`.

**Comment by ChinYikMing** [comment]
> > I’m still confused about file suffixes related to CSV. Could you please provide more details?
> 
> The contents of `*.csv` files are copied from the output of `sha1sum` and are changed to comma-separated manually (the released `sha1sum-*` are still whitespace separated).

So, are the `*.csv` files primarily used to make comparisons easier to view compared to the space-separated `sha1sum-*` files?

**Comment by vacantron** [comment]
> So, are the `*.csv` files primarily used to make comparisons easier to view compared to the space-separated `sha1sum-*` files?

I think both of them are the same with `awk` command. The only difference is the separating character needs to be specified when processing CSV files.

Or, do you care about the `sha1sum-*` files without suffix might be confused with the prebuilt binaries?

**Comment by ChinYikMing** [comment]
> > So, are the `*.csv` files primarily used to make comparisons easier to view compared to the space-separated `sha1sum-*` files?
> 
> I think both of them are the same with `awk` command. The only difference is the separating character needs to be specified when processing CSV files.
> 
> Or, do you care about the `sha1sum-*` files without suffix might be confused with the prebuilt binaries?

I was wondering, since we have the space-separated `sha1sum-*` files, would the `*.csv` files still be necessary?

**Comment by vacantron** [comment]
> I was wondering, since we have the space-separated `sha1sum-*` files, would the `*.csv` files still be necessary?

No, but any file in `rv32emu-prebuilt` won't be updated since it is only for the purpose of releasing prebuilt binaries. We only fetch the files from the RELEASE page.

Or, we can simply clear the repository, but not in this pull request.

**Comment by ChinYikMing** [comment]
> > I was wondering, since we have the space-separated `sha1sum-*` files, would the `*.csv` files still be necessary?
> 
> No, but any file in `rv32emu-prebuilt` won't be updated since it is only for the purpose of releasing prebuilt binaries. We only fetch the files from the RELEASE page.
> 
> Or, we can simply clear the repository, but not in this pull request.

Agree. From my perspective, the `*.csv` files can be a bit confusing in terms of maintainability, especially since the useful ones are available on the RELEASE page, as you mentioned.

**Review by ChinYikMing — APPROVED** [review]
Now, LGTM.

**Pull request merged**